protocols/autonatv2/src/client/handler/dial_request.rs

```diff
@@ -0,0 +1,396 @@
+            return Poll::Ready(ConnectionHandlerEvent::NotifyBehaviour(
+                ToBehaviour::StatusUpdate(status_update),
+            ));
+        }
+        Poll::Pending
+    }
+
+    fn on_behaviour_event(&mut self, event: Self::FromBehaviour) {
+        self.perform_request(event);
+    }
+
+    fn on_connection_event(
+        &mut self,
+        event: ConnectionEvent<
+            Self::InboundProtocol,
+            Self::OutboundProtocol,
+            Self::InboundOpenInfo,
+            Self::OutboundOpenInfo,
+        >,
+    ) {
+        match event {
+            ConnectionEvent::DialUpgradeError(DialUpgradeError { error, .. }) => {
+                tracing::debug!("Dial request failed: {}", error);
+                match self.queued_streams.pop_front() {
+                    Some(stream_tx) => {
+                        let _ = stream_tx.send(Err(error));
+                    }
+                    None => {
+                        tracing::warn!(
+                            "Opened unexpected substream without a pending dial request"
+                        );
+                    }
+                }
+            }
+            ConnectionEvent::FullyNegotiatedOutbound(FullyNegotiatedOutbound {
+                protocol, ..
+            }) => match self.queued_streams.pop_front() {
+                Some(stream_tx) => {
+                    if stream_tx.send(Ok(protocol)).is_err() {
+                        tracing::warn!("Failed to send stream to dead handler");
+                    }
+                }
+                None => {
+                    tracing::warn!("Opened unexpected substream without a pending dial request");
+                }
+            },
+            ConnectionEvent::RemoteProtocolsChange(ProtocolsChange::Added(mut added)) => {
+                if added.any(|p| p.as_ref() == REQUEST_PROTOCOL_NAME) {
+                    self.queued_events
+                        .push_back(ConnectionHandlerEvent::NotifyBehaviour(
+                            ToBehaviour::PeerHasServerSupport,
+                        ));
+                }
+            }
+            _ => {}
+        }
+    }
+}
+
+async fn start_stream_handle(
+    server: PeerId,
+    dial_request: DialRequest,
+    stream_recv: oneshot::Receiver<
+        Result<
+            Stream,
+            StreamUpgradeError<<ReadyUpgrade<StreamProtocol> as OutboundUpgradeSend>::Error>,
+        >,
+    >,
+    mut status_update_tx: mpsc::Sender<Event>,
+) -> Result<TestEnd, crate::client::behaviour::Error> {
+    let substream = match stream_recv.await {
+        Ok(Ok(substream)) => substream,
+        Ok(Err(StreamUpgradeError::Io(io))) => return Err(InternalError::from(io).into()),
+        Ok(Err(StreamUpgradeError::Timeout)) => {
+            return Err(InternalError::Io(io::Error::from(io::ErrorKind::TimedOut)).into())
+        }
+        Ok(Err(StreamUpgradeError::Apply(upgrade_error))) => {
```

Comment:
This should be `Void` if you are using `ReadyUpgrade`.

Reply:
So void::unreachable